### PR TITLE
generic proxy: file access log support

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: generic proxy: file access log support
Additional Description:

Access log support to the generic proxy. See https://github.com/envoyproxy/envoy/issues/28773

Risk Level: low.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

File: source/common/access_log/access_log_impl.h [hunk 1]
```diff
@@ -305,6 +308,48 @@ class AccessLogFactory {
...
+   * Template method to create an access log filter from proto configuration for non-HTTP access
+   * loggers.
+   */
+  template <class Context>
+  static FilterBasePtr<Context>
+  accessLogFilterFromProto(const envoy::config::accesslog::v3::AccessLogFilter& config,
+                           Server::Configuration::CommonFactoryContext& context) {
+    if (!config.has_extension_filter()) {
+      ExceptionUtil::throwEnvoyException(
+          "Access log filter: only extension filter is supported by non-HTTP access loggers.");
```
Comment (soulxu) [inline source/common/access_log/access_log_impl.h:322]: it seems the tcp proxy filter's access logger doesn't reject this.

**Review by soulxu — COMMENTED** [review]
I guess the first question whether we need a separated access logger interface for generic proxy.

File: source/common/access_log/access_log_impl.h (hunk 1 above)
Comment (wbpcode) [reply source/common/access_log/access_log_impl.h:322]: I think tcp proxy use the HTTP access logger directly. It's not the non-HTTP access loggers.

File: envoy/access_log/access_log.h [hunk 2]
```diff
@@ -57,6 +57,35 @@ class AccessLogManager {
...
+public:
+  virtual ~InstanceBase() = default;
+
+  /**
+   * Log a completed request.
+   * @param context supplies the context for the log.
+   * @param stream_info supplies additional information about the request not
+   * contained in the request headers.
+   */
+  virtual void log(const Context& context, const StreamInfo::StreamInfo& stream_info) PURE;
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply envoy/access_log/access_log.h:85]: The original log interface only take stream info and HTTP headers/trailers as the input parameter. This means the original interface couldn't used to log non-HTTP specific field or data. This isn't enough for full-features support of logging.

The new templated interface will be used as alternative of previous interface. And I will deprecated the old interface just like I did to the formatter interface.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**